File: src/lib/package-manager.ts
/**
 * Package Manager Detection and Selection
 * Automatically detects the preferred package manager or lets user choose
 *
 * Supports: npm, pnpm, yarn, bun
 */

import fs from 'fs';
import path from 'path';
import { commandExists, getClaudeDir, readFile, writeFile } from './utils';

/** Configuration for a package manager — commands for install, run, exec, test, build, and dev. */
export interface PackageManagerConfig {
  name: string;
  lockFile: string;
  installCmd: string;
  runCmd: string;
  execCmd: string;
  testCmd: string;
  buildCmd: string;
  devCmd: string;
}

/** Supported package managers with their command configurations. */
export const PACKAGE_MANAGERS: Record<string, PackageManagerConfig> = {
  npm: {
    name: 'npm',
    lockFile: 'package-lock.json',
    installCmd: 'npm install',
    runCmd: 'npm run',
    execCmd: 'npx',
    testCmd: 'npm test',
    buildCmd: 'npm run build',
    devCmd: 'npm run dev'
  },
  pnpm: {
    name: 'pnpm',
    lockFile: 'pnpm-lock.yaml',
    installCmd: 'pnpm install',
    runCmd: 'pnpm',
    execCmd: 'pnpm dlx',
    testCmd: 'pnpm test',
    buildCmd: 'pnpm build',
    devCmd: 'pnpm dev'
  },
  yarn: {
    name: 'yarn',
    lockFile: 'yarn.lock',
    installCmd: 'yarn',
    runCmd: 'yarn',
    execCmd: 'yarn dlx',
    testCmd: 'yarn test',
    buildCmd: 'yarn build',
    devCmd: 'yarn dev'
  },
  bun: {
    name: 'bun',
    lockFile: 'bun.lockb',
    installCmd: 'bun install',
    runCmd: 'bun run',
    execCmd: 'bunx',
    testCmd: 'bun test',
    buildCmd: 'bun run build',
    devCmd: 'bun run dev'
  }
};

/** Priority order for detecting package managers from lock files. */
export const DETECTION_PRIORITY = ['pnpm', 'bun', 'yarn', 'npm'] as const;

function getConfigPath(): string {
  return path.join(getClaudeDir(), 'package-manager.json');
}

interface PackageManagerSavedConfig {
  packageManager?: string;
  setAt?: string;
}

/**
 * Load saved package manager configuration
 */
function loadConfig(): PackageManagerSavedConfig | null {
  const configPath = getConfigPath();
  const content = readFile(configPath);

  if (content) {
    try {
      return JSON.parse(content) as PackageManagerSavedConfig;
    } catch {
      return null;
    }
  }
  return null;
}

/**
 * Save package manager configuration
 */
function saveConfig(config: PackageManagerSavedConfig): void {
  const configPath = getConfigPath();
  writeFile(configPath, JSON.stringify(config, null, 2));
}

/**
 * Detect package manager from lock file in project directory
 */
export function detectFromLockFile(projectDir: string = process.cwd()): string | null {
  for (const pmName of DETECTION_PRIORITY) {
    const pm = PACKAGE_MANAGERS[pmName];
    const lockFilePath = path.join(projectDir, pm.lockFile);

    if (fs.existsSync(lockFilePath)) {
      return pmName;
    }
  }
  return null;
}

/**
 * Detect package manager from package.json packageManager field
 */
export function detectFromPackageJson(projectDir: string = process.cwd()): string | null {
  const packageJsonPath = path.join(projectDir, 'package.json');
  const content = readFile(packageJsonPath);

  if (content) {
    try {
      const pkg = JSON.parse(content) as { packageManager?: string };
      if (pkg.packageManager) {
        const pmName = pkg.packageManager.split('@')[0];
        if (PACKAGE_MANAGERS[pmName]) {
          return pmName;
        }
      }
    } catch {
      // Invalid package.json
    }
  }
  return null;
}

/**
 * Get available package managers (installed on system)
 *
 * WARNING: This spawns child processes for each package manager.
 * Do NOT call during session startup hooks.
 */
export function getAvailablePackageManagers(): string[] {
  const available: string[] = [];

  for (const pmName of Object.keys(PACKAGE_MANAGERS)) {
    if (commandExists(pmName)) {
      available.push(pmName);
    }
  }

  return available;
}

/** Detected or configured package manager with its config and detection source. */
export interface PackageManagerResult {
  name: string;
  config: PackageManagerConfig;
  source: string;
}

/** Options for package manager detection — optional project directory override. */
export interface GetPackageManagerOptions {
  projectDir?: string;
}

/**
 * Get the package manager to use for current project
 *
 * Detection priority:
 * 1. Environment variable CLAUDE_PACKAGE_MANAGER
 * 2. Project-specific config (in .claude/package-manager.json)
 * 3. package.json packageManager field
 * 4. Lock file detection
 * 5. Global user preference (in ~/.claude/package-manager.json)
 * 6. Default to npm
 */
export function getPackageManager(options: GetPackageManagerOptions = {}): PackageManagerResult {
  const { projectDir = process.cwd() } = options;

  // 1. Check environment variable
  const envPm = process.env.CLAUDE_PACKAGE_MANAGER;
  if (envPm && PACKAGE_MANAGERS[envPm]) {
    return { name: envPm, config: PACKAGE_MANAGERS[envPm], source: 'environment' };
  }

  // 2. Check project-specific config
  const projectConfigPath = path.join(projectDir, '.claude', 'package-manager.json');
  const projectConfig = readFile(projectConfigPath);
  if (projectConfig) {
    try {
      const config = JSON.parse(projectConfig) as PackageManagerSavedConfig;
      if (config.packageManager && PACKAGE_MANAGERS[config.packageManager]) {
        return { name: config.packageManager, config: PACKAGE_MANAGERS[config.packageManager], source: 'project-config' };
      }
    } catch {
      // Invalid config
    }
  }

  // 3. Check package.json packageManager field
  const fromPackageJson = detectFromPackageJson(projectDir);
  if (fromPackageJson) {
    return { name: fromPackageJson, config: PACKAGE_MANAGERS[fromPackageJson], source: 'package.json' };
  }

  // 4. Check lock file
  const fromLockFile = detectFromLockFile(projectDir);
  if (fromLockFile) {
    return { name: fromLockFile, config: PACKAGE_MANAGERS[fromLockFile], source: 'lock-file' };
  }

  // 5. Check global user preference
  const globalConfig = loadConfig();
  if (globalConfig?.packageManager && PACKAGE_MANAGERS[globalConfig.packageManager]) {
    return { name: globalConfig.packageManager, config: PACKAGE_MANAGERS[globalConfig.packageManager], source: 'global-config' };
  }

  // 6. Default to npm
  return { name: 'npm', config: PACKAGE_MANAGERS.npm, source: 'default' };
}

/**
 * Set user's preferred package manager (global)
 */
export function setPreferredPackageManager(pmName: string): PackageManagerSavedConfig {
  if (!PACKAGE_MANAGERS[pmName]) {
    throw new Error(`Unknown package manager: ${pmName}`);
  }

  const config: PackageManagerSavedConfig = loadConfig() || {};
  config.packageManager = pmName;
  config.setAt = new Date().toISOString();

  try {
    saveConfig(config);
  } catch (err: unknown) {
    throw new Error(`Failed to save package manager preference: ${(err as Error).message}`);
  }

  return config;
}

/**
 * Set project's preferred package manager
 */
export function setProjectPackageManager(pmName: string, projectDir: string = process.cwd()): PackageManagerSavedConfig {
  if (!PACKAGE_MANAGERS[pmName]) {
    throw new Error(`Unknown package manager: ${pmName}`);
  }

  const configPath = path.join(projectDir, '.claude', 'package-manager.json');

  const config: PackageManagerSavedConfig = {
    packageManager: pmName,
    setAt: new Date().toISOString()
  };

  try {
    writeFile(configPath, JSON.stringify(config, null, 2));
  } catch (err: unknown) {
    throw new Error(`Failed to save package manager config to ${configPath}: ${(err as Error).message}`);
  }
  return config;
}

// Allowed characters in script/binary names
const SAFE_NAME_REGEX = /^[@a-zA-Z0-9_./-]+$/;

/**
 * Get the command to run a script
 */
export function getRunCommand(script: string, options: GetPackageManagerOptions = {}): string {
  if (!script || typeof script !== 'string') {
    throw new Error('Script name must be a non-empty string');
  }
  if (!SAFE_NAME_REGEX.test(script)) {
    throw new Error(`Script name contains unsafe characters: ${script}`);
  }

  const pm = getPackageManager(options);

  switch (script) {
    case 'install':
      return pm.config.installCmd;
    case 'test':
      return pm.config.testCmd;
    case 'build':
      return pm.config.buildCmd;
    case 'dev':
      return pm.config.devCmd;
    default:
      return `${pm.config.runCmd} ${script}`;
  }
}

// Allowed characters in arguments
const SAFE_ARGS_REGEX = /^[@a-zA-Z0-9\s_./:=,'"*+-]+$/;

/**
 * Get the command to execute a package binary
 */
export function getExecCommand(binary: string, args: string = '', options: GetPackageManagerOptions = {}): string {
  if (!binary || typeof binary !== 'string') {
    throw new Error('Binary name must be a non-empty string');
  }
  if (!SAFE_NAME_REGEX.test(binary)) {
    throw new Error(`Binary name contains unsafe characters: ${binary}`);
  }
  if (args && typeof args === 'string' && !SAFE_ARGS_REGEX.test(args)) {
    throw new Error(`Arguments contain unsafe characters: ${args}`);
  }

  const pm = getPackageManager(options);
  return `${pm.config.execCmd} ${binary}${args ? ' ' + args : ''}`;
}

/**
 * Interactive prompt for package manager selection
 */
export function getSelectionPrompt(): string {
  let message = '[PackageManager] No package manager preference detected.\n';
  message += 'Supported package managers: ' + Object.keys(PACKAGE_MANAGERS).join(', ') + '\n';
  message += '\nTo set your preferred package manager:\n';
  message += '  - Global: Set CLAUDE_PACKAGE_MANAGER environment variable\n';
  message += '  - Or add to ~/.claude/package-manager.json: {"packageManager": "pnpm"}\n';
  message += '  - Or add to package.json: {"packageManager": "pnpm@8"}\n';
  message += '  - Or add a lock file to your project (e.g., pnpm-lock.yaml)\n';

  return message;
}

function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Generate a regex pattern that matches commands for all package managers
 */
export function getCommandPattern(action: string): string {
  const patterns: string[] = [];
  const trimmedAction = action.trim();

  if (trimmedAction === 'dev') {
    patterns.push('npm run dev', 'pnpm( run)? dev', 'yarn dev', 'bun run dev');
  } else if (trimmedAction === 'install') {
    patterns.push('npm install', 'pnpm install', 'yarn( install)?', 'bun install');
  } else if (trimmedAction === 'test') {
    patterns.push('npm test', 'pnpm test', 'yarn test', 'bun test');
  } else if (trimmedAction === 'build') {
    patterns.push('npm run build', 'pnpm( run)? build', 'yarn build', 'bun run build');
  } else {
    const escaped = escapeRegex(trimmedAction);
    patterns.push(`npm run ${escaped}`, `pnpm( run)? ${escaped}`, `yarn ${escaped}`, `bun run ${escaped}`);
  }

  return `(${patterns.join('|')})`;
}

File: crates/ecc-domain/src/detection/package_manager.rs
/// Configuration for a package manager.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PackageManagerConfig {
    /// Name of the package manager (e.g., `npm`, `pnpm`).
    pub name: &'static str,
    /// Lock file name for this package manager.
    pub lock_file: &'static str,
    /// Command to install dependencies.
    pub install_cmd: &'static str,
    /// Command prefix to run scripts.
    pub run_cmd: &'static str,
    /// Command to execute a package binary.
    pub exec_cmd: &'static str,
    /// Command to run tests.
    pub test_cmd: &'static str,
    /// Command to build the project.
    pub build_cmd: &'static str,
    /// Command to run the development server.
    pub dev_cmd: &'static str,
}

/// Configuration for npm package manager.
pub const NPM: PackageManagerConfig = PackageManagerConfig {
    name: "npm",
    lock_file: "package-lock.json",
    install_cmd: "npm install",
    run_cmd: "npm run",
    exec_cmd: "npx",
    test_cmd: "npm test",
    build_cmd: "npm run build",
    dev_cmd: "npm run dev",
};

/// Configuration for pnpm package manager.
pub const PNPM: PackageManagerConfig = PackageManagerConfig {
    name: "pnpm",
    lock_file: "pnpm-lock.yaml",
    install_cmd: "pnpm install",
    run_cmd: "pnpm",
    exec_cmd: "pnpm dlx",
    test_cmd: "pnpm test",
    build_cmd: "pnpm build",
    dev_cmd: "pnpm dev",
};

/// Configuration for yarn package manager.
pub const YARN: PackageManagerConfig = PackageManagerConfig {
    name: "yarn",
    lock_file: "yarn.lock",
    install_cmd: "yarn",
    run_cmd: "yarn",
    exec_cmd: "yarn dlx",
    test_cmd: "yarn test",
    build_cmd: "yarn build",
    dev_cmd: "yarn dev",
};

/// Configuration for bun package manager.
pub const BUN: PackageManagerConfig = PackageManagerConfig {
    name: "bun",
    lock_file: "bun.lockb",
    install_cmd: "bun install",
    run_cmd: "bun run",
    exec_cmd: "bunx",
    test_cmd: "bun test",
    build_cmd: "bun run build",
    dev_cmd: "bun run dev",
};

/// All supported package managers.
pub const ALL_CONFIGS: &[&PackageManagerConfig] = &[&NPM, &PNPM, &YARN, &BUN];

/// Priority order for lock file detection (pnpm and bun checked first).
pub const DETECTION_PRIORITY: &[&PackageManagerConfig] = &[&PNPM, &BUN, &YARN, &NPM];

/// Error type for package manager operations.
#[derive(Debug, thiserror::Error)]
pub enum PackageManagerError {
    /// The script name is empty.
    #[error("Script name must be a non-empty string")]
    EmptyScriptName,
    /// The script name contains unsafe characters.
    #[error("Script name contains unsafe characters: {0}")]
    UnsafeScriptName(String),
    /// The arguments contain unsafe characters.
    #[error("Arguments contain unsafe characters: {0}")]
    UnsafeArgs(String),
}

/// Regex for safe script/binary names.
const SAFE_NAME_PATTERN: &str = r"^[@a-zA-Z0-9_./-]+$";

/// Regex for safe command arguments.
const SAFE_ARGS_PATTERN: &str = r#"^[@a-zA-Z0-9\s_./:=,'"*+\-]+$"#;

/// Look up a config by name.
///
/// # Arguments
///
/// * `name` — The package manager name (e.g., `npm`, `pnpm`).
///
/// # Returns
///
/// A reference to the config, or `None` if the name is not recognized.
pub fn find_config(name: &str) -> Option<&'static PackageManagerConfig> {
    ALL_CONFIGS.iter().find(|c| c.name == name).copied()
}

/// Detection source for a package manager result.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum DetectionSource {
    /// Detected from an environment variable.
    Environment,
    /// Detected from a project configuration file.
    ProjectConfig,
    /// Detected from package.json `packageManager` field.
    PackageJson,
    /// Detected from lock file presence.
    LockFile,
    /// Detected from global configuration.
    GlobalConfig,
    /// Using the default package manager.
    Default,
}

impl DetectionSource {
    /// Return the string representation of the detection source.
    ///
    /// # Returns
    ///
    /// A static string describing how the package manager was detected.
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Environment => "environment",
            Self::ProjectConfig => "project-config",
            Self::PackageJson => "package.json",
            Self::LockFile => "lock-file",
            Self::GlobalConfig => "global-config",
            Self::Default => "default",
        }
    }
}

/// Result of detecting the package manager.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PackageManagerResult {
    /// The name of the detected package manager.
    pub name: String,
    /// The configuration for this package manager.
    pub config: &'static PackageManagerConfig,
    /// How the package manager was detected.
    pub source: DetectionSource,
}

use std::sync::LazyLock;

static SAFE_NAME_RE: LazyLock<regex::Regex> = LazyLock::new(|| {
    regex::Regex::new(SAFE_NAME_PATTERN).expect("BUG: invalid SAFE_NAME_PATTERN regex")
});

static SAFE_ARGS_RE: LazyLock<regex::Regex> = LazyLock::new(|| {
    regex::Regex::new(SAFE_ARGS_PATTERN).expect("BUG: invalid SAFE_ARGS_PATTERN regex")
});

/// Validate a script name contains only safe characters.
///
/// # Arguments
///
/// * `name` — The script name to validate.
///
/// # Returns
///
/// `Ok(())` if valid, or `Err` with details about the validation failure.
pub fn validate_script_name(name: &str) -> Result<(), PackageManagerError> {
    if name.is_empty() {
        return Err(PackageManagerError::EmptyScriptName);
    }
    if !SAFE_NAME_RE.is_match(name) {
        return Err(PackageManagerError::UnsafeScriptName(name.to_string()));
    }
    Ok(())
}

/// Validate command arguments contain only safe characters.
///
/// # Arguments
///
/// * `args` — The command arguments to validate.
///
/// # Returns
///
/// `Ok(())` if valid, or `Err` with details about the validation failure.
pub fn validate_args(args: &str) -> Result<(), PackageManagerError> {
    if args.is_empty() {
        return Ok(());
    }
    if !SAFE_ARGS_RE.is_match(args) {
        return Err(PackageManagerError::UnsafeArgs(args.to_string()));
    }
    Ok(())
}

/// Get the command to run a script with the given package manager config.
///
/// # Arguments
///
/// * `config` — The package manager configuration.
/// * `script` — The script name (e.g., `test`, `lint`).
///
/// # Returns
///
/// The full command to run, or an error if the script name is invalid.
pub fn get_run_command(
    config: &PackageManagerConfig,
    script: &str,
) -> Result<String, PackageManagerError> {
    validate_script_name(script)?;

    Ok(match script {
        "install" => config.install_cmd.to_string(),
        "test" => config.test_cmd.to_string(),
        "build" => config.build_cmd.to_string(),
        "dev" => config.dev_cmd.to_string(),
        _ => format!("{} {script}", config.run_cmd),
    })
}

/// Get the command to execute a package binary.
///
/// # Arguments
///
/// * `config` — The package manager configuration.
/// * `binary` — The binary name to execute (e.g., `jest`).
/// * `args` — Optional arguments to pass to the binary.
///
/// # Returns
///
/// The full command to run, or an error if the binary or args are invalid.
pub fn get_exec_command(
    config: &PackageManagerConfig,
    binary: &str,
    args: Option<&str>,
) -> Result<String, PackageManagerError> {
    validate_script_name(binary)?;
    if let Some(a) = args {
        validate_args(a)?;
    }

    let base = format!("{} {binary}", config.exec_cmd);
    Ok(match args {
        Some(a) if !a.is_empty() => format!("{base} {a}"),
        _ => base,
    })
}

/// Escape regex special characters in a string.
///
/// # Arguments
///
/// * `s` — The string to escape.
///
/// # Returns
///
/// The escaped string safe for use in a regex pattern.
fn escape_regex(s: &str) -> String {
    let special = r".*+?^${}()|[]\";
    let mut escaped = String::with_capacity(s.len() * 2);
    for ch in s.chars() {
        if special.contains(ch) {
            escaped.push('\\');
        }
        escaped.push(ch);
    }
    escaped
}

/// Generate a regex pattern that matches commands for all package managers.
///
/// # Arguments
///
/// * `action` — The action (e.g., `dev`, `install`, `test`, or a custom script name).
///
/// # Returns
///
/// A regex pattern that matches how this action is invoked in all supported package managers.
pub fn get_command_pattern(action: &str) -> String {
    let trimmed = action.trim();
    let patterns = match trimmed {
        "dev" => vec![
            "npm run dev".to_string(),
            "pnpm( run)? dev".to_string(),
            "yarn dev".to_string(),
            "bun run dev".to_string(),
        ],
        "install" => vec![
            "npm install".to_string(),
            "pnpm install".to_string(),
            "yarn( install)?".to_string(),
            "bun install".to_string(),
        ],
        "test" => vec![
            "npm test".to_string(),
            "pnpm test".to_string(),
            "yarn test".to_string(),
            "bun test".to_string(),
        ],
        "build" => vec![
            "npm run build".to_string(),
            "pnpm( run)? build".to_string(),
            "yarn build".to_string(),
            "bun run build".to_string(),
        ],
        _ => {
            let escaped = escape_regex(trimmed);
            vec![
                format!("npm run {escaped}"),
                format!("pnpm( run)? {escaped}"),
                format!("yarn {escaped}"),
                format!("bun run {escaped}"),
            ]
        }
    };

    format!("({})", patterns.join("|"))
}

#[cfg(test)]
mod tests {
    use super::*;

    // --- Config constants ---

    #[test]
    fn npm_config() {
        assert_eq!(NPM.name, "npm");
        assert_eq!(NPM.lock_file, "package-lock.json");
        assert_eq!(NPM.install_cmd, "npm install");
        assert_eq!(NPM.exec_cmd, "npx");
    }

    #[test]
    fn pnpm_config() {
        assert_eq!(PNPM.name, "pnpm");
        assert_eq!(PNPM.lock_file, "pnpm-lock.yaml");
        assert_eq!(PNPM.run_cmd, "pnpm");
        assert_eq!(PNPM.exec_cmd, "pnpm dlx");
    }

    #[test]
    fn yarn_config() {
        assert_eq!(YARN.name, "yarn");
        assert_eq!(YARN.lock_file, "yarn.lock");
        assert_eq!(YARN.install_cmd, "yarn");
        assert_eq!(YARN.exec_cmd, "yarn dlx");
    }

    #[test]
    fn bun_config() {
        assert_eq!(BUN.name, "bun");
        assert_eq!(BUN.lock_file, "bun.lockb");
        assert_eq!(BUN.exec_cmd, "bunx");
    }

    #[test]
    fn all_configs_has_four() {
        assert_eq!(ALL_CONFIGS.len(), 4);
    }

    #[test]
    fn detection_priority_pnpm_first() {
        assert_eq!(DETECTION_PRIORITY[0].name, "pnpm");
    }

    // --- find_config ---

    #[test]
    fn find_config_npm() {
        assert_eq!(find_config("npm").unwrap().name, "npm");
    }

    #[test]
    fn find_config_unknown() {
        assert!(find_config("unknown").is_none());
    }

    // --- validate_script_name ---

    #[test]
    fn validate_name_valid() {
        assert!(validate_script_name("test").is_ok());
        assert!(validate_script_name("@scope/pkg").is_ok());
        assert!(validate_script_name("my-script").is_ok());
    }

    #[test]
    fn validate_name_empty() {
        assert!(validate_script_name("").is_err());
    }

    #[test]
    fn validate_name_unsafe() {
        assert!(validate_script_name("rm -rf /").is_err());
        assert!(validate_script_name("$(whoami)").is_err());
    }

    // --- validate_args ---

    #[test]
    fn validate_args_valid() {
        assert!(validate_args("--port=3000").is_ok());
        assert!(validate_args("--config 'test.json'").is_ok());
    }

    #[test]
    fn validate_args_empty() {
        assert!(validate_args("").is_ok());
    }

    #[test]
    fn validate_args_unsafe() {
        assert!(validate_args("$(whoami)").is_err());
        assert!(validate_args("; rm -rf /").is_err());
    }

    // --- get_run_command ---

    #[test]
    fn run_command_install() {
        assert_eq!(get_run_command(&NPM, "install").unwrap(), "npm install");
        assert_eq!(get_run_command(&YARN, "install").unwrap(), "yarn");
    }

    #[test]
    fn run_command_test() {
        assert_eq!(get_run_command(&NPM, "test").unwrap(), "npm test");
        assert_eq!(get_run_command(&PNPM, "test").unwrap(), "pnpm test");
    }

    #[test]
    fn run_command_build() {
        assert_eq!(get_run_command(&NPM, "build").unwrap(), "npm run build");
        assert_eq!(get_run_command(&PNPM, "build").unwrap(), "pnpm build");
    }

    #[test]
    fn run_command_dev() {
        assert_eq!(get_run_command(&NPM, "dev").unwrap(), "npm run dev");
        assert_eq!(get_run_command(&YARN, "dev").unwrap(), "yarn dev");
    }

    #[test]
    fn run_command_custom() {
        assert_eq!(get_run_command(&NPM, "lint").unwrap(), "npm run lint");
        assert_eq!(get_run_command(&PNPM, "lint").unwrap(), "pnpm lint");
        assert_eq!(get_run_command(&YARN, "lint").unwrap(), "yarn lint");
        assert_eq!(get_run_command(&BUN, "lint").unwrap(), "bun run lint");
    }

    #[test]
    fn run_command_empty_name() {
        assert!(get_run_command(&NPM, "").is_err());
    }

    #[test]
    fn run_command_unsafe_name() {
        assert!(get_run_command(&NPM, "$(whoami)").is_err());
    }

    // --- get_exec_command ---

    #[test]
    fn exec_command_npm() {
        assert_eq!(get_exec_command(&NPM, "jest", None).unwrap(), "npx jest");
    }

    #[test]
    fn exec_command_with_args() {
        assert_eq!(
            get_exec_command(&NPM, "jest", Some("--watch")).unwrap(),
            "npx jest --watch"
        );
    }

    #[test]
    fn exec_command_pnpm() {
        assert_eq!(
            get_exec_command(&PNPM, "create-react-app", Some("my-app")).unwrap(),
            "pnpm dlx create-react-app my-app"
        );
    }

    #[test]
    fn exec_command_bun() {
        assert_eq!(
            get_exec_command(&BUN, "vitest", None).unwrap(),
            "bunx vitest"
        );
    }

    #[test]
    fn exec_command_empty_binary() {
        assert!(get_exec_command(&NPM, "", None).is_err());
    }

    #[test]
    fn exec_command_unsafe_args() {
        assert!(get_exec_command(&NPM, "jest", Some("; rm -rf /")).is_err());
    }

    // --- get_command_pattern ---

    #[test]
    fn command_pattern_dev() {
        let pattern = get_command_pattern("dev");
        assert!(pattern.contains("npm run dev"));
        assert!(pattern.contains("pnpm( run)? dev"));
        assert!(pattern.contains("yarn dev"));
        assert!(pattern.contains("bun run dev"));
    }

    #[test]
    fn command_pattern_install() {
        let pattern = get_command_pattern("install");
        assert!(pattern.contains("npm install"));
        assert!(pattern.contains("yarn( install)?"));
    }

    #[test]
    fn command_pattern_test() {
        let pattern = get_command_pattern("test");
        assert!(pattern.contains("npm test"));
        assert!(pattern.contains("bun test"));
    }

    #[test]
    fn command_pattern_build() {
        let pattern = get_command_pattern("build");
        assert!(pattern.contains("npm run build"));
        assert!(pattern.contains("pnpm( run)? build"));
    }

    #[test]
    fn command_pattern_custom() {
        let pattern = get_command_pattern("lint");
        assert!(pattern.contains("npm run lint"));
        assert!(pattern.contains("pnpm( run)? lint"));
        assert!(pattern.contains("yarn lint"));
        assert!(pattern.contains("bun run lint"));
    }

    #[test]
    fn command_pattern_trimmed() {
        let pattern = get_command_pattern("  dev  ");
        assert!(pattern.contains("npm run dev"));
    }

    #[test]
    fn command_pattern_regex_escaping() {
        let pattern = get_command_pattern("test.spec");
        // The dot should be escaped
        assert!(pattern.contains(r"test\.spec"));
    }

    // --- DetectionSource ---

    #[test]
    fn detection_source_as_str() {
        assert_eq!(DetectionSource::Environment.as_str(), "environment");
        assert_eq!(DetectionSource::LockFile.as_str(), "lock-file");
        assert_eq!(DetectionSource::Default.as_str(), "default");
    }

    // --- PackageManagerError ---

    /// PC-006: PackageManagerError enum variants compile and are accessible.
    #[test]
    fn package_manager_error_empty_name_is_debug() {
        let err = PackageManagerError::EmptyScriptName;
        let s = format!("{err:?}");
        assert!(!s.is_empty());
    }

    #[test]
    fn package_manager_error_unsafe_name_carries_message() {
        let err = PackageManagerError::UnsafeScriptName("$(whoami)".to_string());
        let s = err.to_string();
        assert!(s.contains("$(whoami)"), "expected name in error, got: {s}");
    }

    #[test]
    fn package_manager_error_unsafe_args_carries_message() {
        let err = PackageManagerError::UnsafeArgs("; rm -rf /".to_string());
        let s = err.to_string();
        assert!(s.contains("; rm"), "expected args in error, got: {s}");
    }
}
